Objects/typeobject.c

```diff
@@ -3741,6 +3741,91 @@ type_is_gc(PyTypeObject *type)
     return type->tp_flags & Py_TPFLAGS_HEAPTYPE;
 }
 
+
+// TODO: MM: Move the following three functions?
+static int
+is_typing_name(PyObject *obj, char *name)
+{
+    PyTypeObject *type = Py_TYPE(obj);
+    if (strcmp(type->tp_name, name) != 0) {
+        return 0;
+    }
+    PyObject *module = PyObject_GetAttrString((PyObject *)type, "__module__");
+    if (module == NULL) {
+        return -1;
+    }
+    int res = PyUnicode_Check(module)
+        && _PyUnicode_EqualToASCIIString(module, "typing");
+    Py_DECREF(module);
+    return res;
+}
+
+static int
+is_typevar(PyObject *obj)
+{
+    return is_typing_name(obj, "TypeVar");
+}
+
+static int
+is_genericalias(PyObject *obj)
+{
+    return is_typing_name(obj, "GenericAlias");
+}
+
+static int
+is_new_type(PyObject *obj)
+{
+    if (!PyObject_IsInstance((PyObject *)obj, (PyObject *)&PyFunction_Type)) {
+        return 0;
+    }
+    PyObject *module = PyObject_GetAttrString((PyObject *)obj, "__module__");
+    if (module == NULL) {
+        return 0;
+    }
+    return PyUnicode_Check(module) && _PyUnicode_EqualToASCIIString(module, "typing");
+}
+
+static int
+is_not_unionable (PyObject *obj)
+{
+    return (obj != Py_None) &
+        !is_genericalias(obj) &&
+        !is_typevar(obj) &&
+        !is_new_type(obj) &&
+        (PyObject_IsInstance(obj, (PyObject *)&PyType_Type) != 1) &&
+        (PyObject_IsInstance(obj, (PyObject *)&Py_UnionType) != 1);
+}
+
+static PyObject *
+type_or(PyTypeObject* self, PyObject* param) {
+    // Check param is a PyType or GenericAlias
+    if ((param == NULL) || is_not_unionable(param)  || is_not_unionable(self))
```

Comment:
```suggestion
    if ((param == NULL) || is_not_unionable((PyObject *)param)  || is_not_unionable(self))
```
Fixes the error reported by GitHub Actions below.